cfep-9-migrator-file.md

```diff
@@ -0,0 +1,209 @@
+# @mariusvniekerk's proposal
+
+This is a general sketch of what I'd like migrators to be able to achieve.  
+Effectively this means we can decompose all of our migrations into a small set of primitives 
+that we can apply to generate a resulting desired target.
+
+----
+
+Given a feedstock with the following pinnings
+
+```yaml
+# Existing pins
+a:
+  - 1.5
+b:
+  - 1.2
+someflag:
+  - disabled
+```
+
+There are four classes of pinning we can do in general.
+
+## 1. Simple Additive Version Migrations
+
+<table>
+<tr>
+<td>
+
+```yaml
+# Existing pins
+a:
+  - 1.5
+b:
+  - 1.2
+someflag:
+  - disabled
+```
+</td>
+<td> with </td>
+<td>
+
+```yaml
+# migrator
+__migrator: kind: version
+a:
+  - 1.6
+b:
+  - 1.1
+```
+</td>
+<td> -> </td>
+<td>
+
+```yaml
+# resulting pins
+a:
+  - 1.6
+b:
+  - 1.2
+someflag:
+  - disabled
+```
+</td>
+</tr>
+</table>
+
+These are the most common version migrations and encompass things like moving the pinnings for openssl etc.
+
+Note that the version of `b` was NOT decreased.  This is intentional as we don't wish to downgrade accidentally.  In practice many migrators could be in flight and may affect the same version.  In this case the higher version should always 
+win.  If we want to downgrade there is a mechanism but it is purposefully clunky and bad.
+
+## 2. Matrix Version Addition changes
+
+<table>
+<tr>
+<td>
+
+```yaml
+# Existing pins
+a:
+  - 1.5
+python:
+  - 3.6
+someflag:
+  - disabled
+```
+</td>
+<td> with </td>
+<td>
+
+```yaml
+# migrator
+__migrator: kind: version
+python:
+  - 3.6
+  - 3.8a1
+```
+</td>
+<td> -> </td>
+<td>
+
+```yaml
+# resulting pins
+a:
+  - 1.5
+b:
+  - 3.6
+  - 3.8a1
+someflag:
+  - disabled
+```
+</td>
+</tr>
+</table>
+
+These are things like the compiler migration and can be thought of as additions to a version.  
+These need substantial care when dealing with things like zip_keys.  But in general it should be
+feasible to perform for many common cases.
```

Comment:
I’m concerned we are biting off more than we can chew here. The 80% solution of building automated tools for simpler migrations but avoiding those tools for complicated migrations seems prudent. In other words, for a complicated migration we want to have a human manually code the migration. My idea of the goal here is to have the automated migrations cover only the simplest cases where they can be effective and have a low chance of failure.

Reply:
which is mostly why i'm pointing out zip_keys being a problem

Reply:
I don't see how code that performs the operations above (with selectors which are not really adressed) really helps anything though. You are suggesting that we build and maintain code for basically what git can do for us in terms of diffs. We have to exclude complicated migrations in either case.

Did I miss something?

Reply:
I think the main issue here is that git diffs require a set order to be run in. We would like to be able to apply these diffs in any order, since more than one migration could be running at a time and we don't want to block if possible.